pkg/referrer/referrer.go

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2023. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package referrer
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/containerd/nydus-snapshotter/pkg/auth"
+	"github.com/containerd/nydus-snapshotter/pkg/label"
+
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+const metadataNameInLayer = "image/image.boot"
+
+type referrer struct {
+	// The resolver is used for image pull or fetches requests. The best practice
+	// in containerd is that each resolver instance is used only once for a request
+	// and is destroyed when the request completes. When a registry token expires,
+	// the resolver does not re-apply for a new token, so it's better to create a
+	// new resolver instance using resolverFunc for each request.
+	resolverFunc  func(plainHTTP bool) remotes.Resolver
+	WithPlainHTTP bool
+}
+
+func newReferrer(keyChain *auth.PassKeyChain, insecure bool) *referrer {
+	// nolint:unparam
+	credFunc := func(string) (string, string, error) {
+		if keyChain == nil {
+			return "", "", nil
+		}
+		return keyChain.Username, keyChain.Password, nil
+	}
+
+	newClient := func(insecure bool) *http.Client {
+		client := http.DefaultClient
+		transport := http.DefaultTransport.(*http.Transport)
+		transport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: insecure,
+		}
+		client.Transport = transport
+		return client
+	}
+
+	resolverFunc := func(plainHTTP bool) remotes.Resolver {
+		registryHosts := docker.ConfigureDefaultRegistries(
+			docker.WithAuthorizer(
+				docker.NewDockerAuthorizer(
+					docker.WithAuthClient(newClient(insecure)),
+					docker.WithAuthCreds(credFunc),
+				),
+			),
+			docker.WithClient(newClient(insecure)),
+			docker.WithPlainHTTP(func(host string) (bool, error) {
+				return plainHTTP, nil
+			}),
+		)
+
+		return docker.NewResolver(docker.ResolverOptions{
+			Hosts: registryHosts,
+		})
+	}
+
+	return &referrer{
+		resolverFunc:  resolverFunc,
+		WithPlainHTTP: true,
+	}
+}
+
+// checkReferrer fetches the referrers and parse out the nydus
+// image by specified manifest digest.
+// it's using distribution list referrers API.
+func (r *referrer) checkReferrer(ctx context.Context, ref string, manifestDigest digest.Digest) (*ocispec.Descriptor, error) {
```

Comment:
It's better to make this a common OCI artifact downloading capability, not only for downloading referenced artifacts.

Reply:
Fixed, moved to `pkg/remote`.